node/libs/storage/src/block_store.rs

```diff
@@ -0,0 +1,218 @@
+        self.first.header().number <= number && number <= self.last.header().number
+    }
+
+    /// Number of the next block that can be stored in the `BlockStore`.
+    /// (i.e. `last` + 1).
+    pub fn next(&self) -> validator::BlockNumber {
+        self.last.header().number.next()
+    }
+}
+
+/// Storage of a continuous range of L2 blocks.
+///
+/// Implementations **must** propagate context cancellation using [`StorageError::Canceled`].
+#[async_trait::async_trait]
+pub trait PersistentBlockStore: fmt::Debug + Send + Sync {
+    /// Range of blocks avaliable in storage. None iff storage is empty.
+    /// Consensus code calls this method only once and then tracks the
+    /// range of avaliable blocks internally.
+    async fn state(&self, ctx: &ctx::Ctx) -> ctx::Result<Option<BlockStoreState>>;
+
+    /// Gets a block by its number. Returns None if block is missing.
+    async fn block(
+        &self,
+        ctx: &ctx::Ctx,
+        number: validator::BlockNumber,
+    ) -> ctx::Result<Option<validator::FinalBlock>>;
+
+    /// Persistently store a block.
+    /// Implementations are only required to accept a block directly after the current last block,
+    /// so that the stored blocks always constitute a continuous range.
+    /// Implementation should return only after the block is stored PERSISTENTLY -
+    /// consensus liveness property depends on this behavior.
+    async fn store_next_block(
+        &self,
+        ctx: &ctx::Ctx,
+        block: &validator::FinalBlock,
+    ) -> ctx::Result<()>;
+}
+
+#[derive(Debug)]
+struct Inner {
+    inmem: sync::watch::Sender<BlockStoreState>,
+    persisted: BlockStoreState,
+    // TODO: this data structure can be optimized to a VecDeque (or even just a cyclical buffer).
+    cache: BTreeMap<validator::BlockNumber, validator::FinalBlock>,
+    cache_capacity: usize,
+}
+
+/// A wrapper around a PersistentBlockStore which adds caching blocks in-memory
+/// and other useful utilities.
+#[derive(Debug)]
+pub struct BlockStore {
+    inner: sync::watch::Sender<Inner>,
+    persistent: Box<dyn PersistentBlockStore>,
+}
+
+/// Runner of the BlockStore background tasks.
+#[must_use]
+pub struct BlockStoreRunner(Arc<BlockStore>);
+
+impl BlockStoreRunner {
+    /// Runs the background tasks of the BlockStore.
+    pub async fn run(self, ctx: &ctx::Ctx) -> anyhow::Result<()> {
+        let res = async {
+            let inner = &mut self.0.inner.subscribe();
+            loop {
+                let block = sync::wait_for(ctx, inner, |inner| !inner.cache.is_empty())
+                    .await?
+                    .cache
+                    .first_key_value()
+                    .unwrap()
+                    .1
+                    .clone();
+                self.0.persistent.store_next_block(ctx, &block).await?;
+                self.0.inner.send_modify(|inner| {
+                    debug_assert!(inner.persisted.next() == block.header().number);
```

Comment:
Nit: You can use `debug_assert_eq` here for more expressive panic messages.

Reply:
done